### ISSUE TITLE
Docker Compose ?

### ISSUE THREAD

**Issue by r3mi** [testcontainers/testcontainers-node#71]
Hello,

is there any way to use with a set of services, using a docker-compose.yml ? or is there any plan to add this functionality ?

thanks for testcontainers-node !

**Comment by polaroi8d** [comment]
Hello @r3mi, 

You could build an image from docker-compose and use this image to start a new GenericContainer.

**Comment by r3mi** [comment]
Hello @polaroi8d , 
Thanks. 
I am not sure how that would work. As I described, the `docker-compose.yml` contains a set of services (e.g. a backend and its database dependency). It is encapsulated as a whole (from a network or volume point of view). Do you foresee a way to make it work with a GenericContainer ?

**Comment by jayarjo** [comment]
TestContainers Java seems to have that: https://www.testcontainers.org/modules/docker_compose/. So this one should have as well at some point future. From what I know there is no support for that at the moment. And author seems to be busy doing other things. Consider a PR maybe.

**Comment by kezhenxu94** [comment]
Really in need of this feature, is anyone working on this? If not, I can help on this maybe

**Comment by jayarjo** [comment]
I will be needing this feature in my app, so I might take a look into it. @cristianrgreco what would be your ideas/suggestions about it?

**Comment by cristianrgreco** [comment]
I haven't really thought about this yet, was kind of hoping it wouldn't be required :smile: 

@kezhenxu94, @jayarjo recently committed creating a network, which would allow communication between containers, would that help you for now?

@jayarjo I guess the benefit of implementing docker-compose would be to export the GenericContainer definitions to a docker-compose.yml? What other benefits are there?

**Comment by kezhenxu94** [comment]
> @kezhenxu94, @jayarjo recently committed creating a network, which would allow communication between containers, would that help you for now?

That helps for sure, but as I'm building test framework for many developers, it would be more helpful that it supports docker-compose:

1. Less codes: I used health check and service dependencies to control the start up order in `docker-compose.yml`, no need to orchestrate the services in codes;
1. Reusable: I have some common service definitions that are to be inherited in specific test cases, with `docker-compose.yml`'s extend ability, no codes are needed;
1. Standalone: the `docker-compose.yml` can be used to start up all the services for manual tests, for example, we have to do some manually test in the release process, starting up from the `docker-compose.yml` is faster and we can reuse the services orchestration that are used in the test, which is also convenient for debugging, i.e. I can orchestrate the services before writing any line of code, and simply start up all services and write the verification codes.

Other implementations like Java and Python all support `docker-compose.yml`, don't know why it's not the case in NodeJS.

**Comment by jayarjo** [comment]
I personally think that it is another must have kind of thing. @kezhenxu94 if you have time to work on it, please do. I can also help.

**Comment by cristianrgreco** [comment]
FYI I have started work on this, see the `docker-compose` branch. I will share API info as soon as I have it

**Comment by cristianrgreco** [comment]
I've raised a PR in case anyone wants to have a look https://github.com/testcontainers/testcontainers-node/pull/94. Here's an excerpt from the README:

## Docker Compose

Testcontainers supports docker-compose. For example for the following `docker-compose.yml`:

```yaml
version: "3"
services:
  redis:
    image: redis:latest
    ports:
      - 6379
  postgres:
    image: postgres:latest
    ports:
      - 5432
```

You can start and stop the environment, and interact with its containers.

```javascript
const path = require("path");
const redis = require("async-redis");
const { DockerComposeEnvironment } = require("testcontainers");

(async () => {
  const composeFilePath = path.resolve(__dirname, "dir-containing-docker-compose-yml");
  const composeFile = "docker-compose.yml"
  const environment = await new DockerComposeEnvironment(composeFilePath, composeFile).up();
  const container = environment.getContainer("redis_1");
  
  const redisClient = redis.createClient(
    container.getMappedPort(6379),
    container.getContainerIpAddress(),
  );
  await redisClient.quit();
  await environment.down();
})();
```

Create the containers with their own wait strategies:

```javascript
const { DockerComposeEnvironment } = require("testcontainers");

const environment = await new DockerComposeEnvironment(composeFilePath, composeFile)
  .withWaitStrategy("redis_1", Wait.forLogMessage("Ready to accept connections"))
  .withWaitStrategy("postgres_1", Wait.forHealthCheck())
  .up();
```

Once the environment has started, you can interact with the containers as you would any other `GenericContainer`:

```javascript
const { DockerComposeEnvironment } = require("testcontainers");

const environment = await new DockerComposeEnvironment(composeFilePath, composeFile).up();

const container = environment.getContainer("alpine_1");
const { output, exitCode } = await container.exec(["echo", "hello", "world"]);
```

**Comment by cristianrgreco** [comment]
Thanks for raising this, it has been released in 2.17.0. Please raise any issues/feature requests as new issues :slightly_smiling_face:
